contracts/protocol/libs/FundsLib.sol

```diff
@@ -202,29 +201,134 @@ library FundsLib {
             }
         }
 
-        // Store payoffs to availablefunds and notify the external observers
         address exchangeToken = offer.exchangeToken;
-        uint256 sellerId = offer.sellerId;
-        uint256 buyerId = exchange.buyerId;
+
+        // Original seller and last buyer are done
+        // Release funds to intermediate sellers (if they exist)
+        // and add the protocol fee to the total
+        {
+            (uint256 sequentialProtocolFee, uint256 sequentialRoyalties) = releaseFundsToIntermediateSellers(
+                _exchangeId,
+                exchange.state,
+                price,
+                exchangeToken
+            );
+            sellerPayoff += sequentialRoyalties; // revisit this
+            protocolFee += sequentialProtocolFee;
+        }
+
+        // Store payoffs to availablefunds and notify the external observers
         address sender = EIP712Lib.msgSender();
         if (sellerPayoff > 0) {
-            increaseAvailableFunds(sellerId, exchangeToken, sellerPayoff);
-            emit FundsReleased(_exchangeId, sellerId, exchangeToken, sellerPayoff, sender);
+            increaseAvailableFundsAndEmitEvent(_exchangeId, offer.sellerId, exchangeToken, sellerPayoff, sender);
         }
         if (buyerPayoff > 0) {
-            increaseAvailableFunds(buyerId, exchangeToken, buyerPayoff);
-            emit FundsReleased(_exchangeId, buyerId, exchangeToken, buyerPayoff, sender);
+            increaseAvailableFundsAndEmitEvent(_exchangeId, exchange.buyerId, exchangeToken, buyerPayoff, sender);
         }
+
         if (protocolFee > 0) {
             increaseAvailableFunds(0, exchangeToken, protocolFee);
             emit ProtocolFeeCollected(_exchangeId, exchangeToken, protocolFee, sender);
         }
         if (agentFee > 0) {
             // Get the agent for offer
             uint256 agentId = ProtocolLib.protocolLookups().agentIdByOffer[exchange.offerId];
-            increaseAvailableFunds(agentId, exchangeToken, agentFee);
-            emit FundsReleased(_exchangeId, agentId, exchangeToken, agentFee, sender);
+            increaseAvailableFundsAndEmitEvent(_exchangeId, agentId, exchangeToken, agentFee, sender);
+        }
+    }
+
+    function releaseFundsToIntermediateSellers(
```

Comment:
Could you please add natspec comments to this function?

Reply:
Addded